kimchi/src/circuits/polynomials/foreign_field_add/gadget.rs

```diff
@@ -0,0 +1,440 @@
+        // Copy bound_mi -> Next(4)
+        circuit_gates.connect_cell_pair((bound_row + 1, 0), (final_row, 4));
+        // Copy bound_hi -> Next(5)
+        circuit_gates.connect_cell_pair((bound_row + 2, 0), (final_row, 5));
+
+        (start_row + circuit_gates.len(), circuit_gates)
+    }
+
+    /// Verifies the foreign field addition gadget
+    pub fn verify_foreign_field_add<G: KimchiCurve<ScalarField = F>>(
+        &self,
+        _: usize,
+        witness: &[Vec<F>; COLUMNS],
+        cs: &ConstraintSystem<F>,
+    ) -> CircuitGateResult<()> {
+        if !circuit_gates().contains(&self.typ) {
+            return Err(CircuitGateError::InvalidCircuitGateType(self.typ));
+        }
+
+        // Pad the witness to domain d1 size
+        let padding_length = cs
+            .domain
+            .d1
+            .size
+            .checked_sub(witness[0].len() as u64)
+            .unwrap();
+        let mut witness = witness.clone();
+        for w in &mut witness {
+            w.extend(std::iter::repeat(F::zero()).take(padding_length as usize));
+        }
+
+        // Compute witness polynomial
+        let witness_poly: [DensePolynomial<F>; COLUMNS] = array::from_fn(|i| {
+            Evaluations::<F, D<F>>::from_vec_and_domain(witness[i].clone(), cs.domain.d1)
+                .interpolate()
+        });
+
+        // Compute permutation polynomial
+        let rng = &mut StdRng::from_seed([0u8; 32]);
+        let beta = F::rand(rng);
+        let gamma = F::rand(rng);
+        let z_poly = cs
+            .perm_aggreg(&witness, &beta, &gamma, rng)
+            .map_err(|_| CircuitGateError::InvalidCopyConstraint(self.typ))?;
+
+        // Compute witness polynomial evaluations
+        let witness_evals = cs.evaluate(&witness_poly, &z_poly);
+
+        let mut index_evals = HashMap::new();
+        index_evals.insert(
+            self.typ,
+            &cs.foreign_field_add_selector_polys.as_ref().unwrap()
+                [circuit_gate_selector_index(self.typ)]
+            .eval8,
+        );
+
+        // Set up lookup environment
+        let lcs = cs
+            .lookup_constraint_system
+            .as_ref()
+            .ok_or(CircuitGateError::MissingLookupConstraintSystem(self.typ))?;
+
+        let lookup_env_data = set_up_lookup_env_data(
+            self.typ,
+            cs,
+            &witness,
+            &beta,
+            &gamma,
+            &lcs.configuration.lookup_info,
+        )?;
+        let lookup_env = Some(LookupEnvironment {
+            aggreg: &lookup_env_data.aggreg8,
+            sorted: &lookup_env_data.sorted8,
+            selectors: &lcs.lookup_selectors,
+            table: &lookup_env_data.joint_lookup_table_d8,
+            runtime_selector: None,
+            runtime_table: None,
+        });
+
+        // Initialize the foreign field modulus constant
+        let foreign_field_modulus = cs.foreign_field_modulus.clone();
```

Comment:
This variable is used in only one place, so remove it and update as described below.

Reply:
oh I see right